src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs

```diff
@@ -145,23 +145,21 @@ public void Dispose()
             {
                 IsDisposed = true;
 
-                // Ensure the _scope reference is cleared because it can reference other CacheEntry instances.
-                // This CacheEntry is going to be put into a MemoryCache, and we don't want to root unnecessary objects.
-                _scope.Dispose();
-                _scope = null;
-
                 // Don't commit or propagate options if the CacheEntry Value was never set.
                 // We assume an exception occurred causing the caller to not set the Value successfully,
                 // so don't use this entry.
                 if (IsValueSet)
                 {
                     _cache.SetEntry(this);
 
-                    if (CanPropagateOptions())
+                    if (_previous != null && CanPropagateOptions())
                     {
-                        PropagateOptions(CacheEntryHelper.Current);
+                        PropagateOptions(_previous);
                     }
                 }
+
+                CacheEntryHelper.ExitScope(this, _previous);
```

Comment:
Is there any way changing the ordering between popping the "scope" off the stack and calling `_cache.SetEntry` could break anything? Before the scope was removed before calling `_cache.SetEntry`. Now it happens after.

Reply:
So far, when `Dispose` was called, we were first setting the `CacheEntryHelper.Current` to `previous` and then accessing `CacheEntryHelper.Current` again to get it. With my change, we take advantage of knowing "previous" and resetting it after using its value (one access to async local instead of two).

I don't think that this reordering can break anything.